Python provider not used for `pants package` of `pex_binary` target

**Describe the bug**
When using `pants.backend.python.providers.experimental.pyenv` test, lint and check works fine, but package fails.

**Pants version**
2.21.0

**OS**
Ubuntu

**Additional info**

```
13:03:09  pants lint check test package ::
13:03:15  13:03:15.48 [INFO] waiting for pantsd to start...
13:03:18  13:03:17.78 [INFO] pantsd started
13:03:18  13:03:18.30 [INFO] Initializing scheduler...
13:03:24  13:03:24.02 [INFO] Scheduler initialized.
13:03:24  13:03:24.51 [INFO] Starting: Choose specific version for Python 3.9
13:03:24  13:03:24.68 [INFO] Completed: Choose specific version for Python 3.9
13:03:24  13:03:24.92 [INFO] Starting: Building isort.pex from resource://pants.backend.python.lint.isort/isort.lock
13:03:24  13:03:24.92 [INFO] Starting: Building black.pex from resource://pants.backend.python.lint.black/black.lock
13:03:25  13:03:24.97 [INFO] Starting: Building autoflake.pex from resource://pants.backend.python.lint.autoflake/autoflake.lock
13:03:25  13:03:24.97 [INFO] Starting: Building docformatter.pex from resource://pants.backend.python.lint.docformatter/docformatter.lock
13:03:25  13:03:24.97 [INFO] Starting: Building flake8.pex from resource://pants.backend.python.lint.flake8/flake8.lock
13:03:25  13:03:25.01 [INFO] Canceled: Building black.pex from resource://pants.backend.python.lint.black/black.lock
13:03:25  13:03:25.01 [INFO] Starting: Building black.pex from resource://pants.backend.python.lint.black/black.lock
13:03:25  13:03:25.01 [INFO] Canceled: Building black.pex from resource://pants.backend.python.lint.black/black.lock
13:03:25  13:03:25.01 [INFO] Starting: Building black.pex from resource://pants.backend.python.lint.black/black.lock
13:03:27  13:03:26.58 [INFO] Completed: Building isort.pex from resource://pants.backend.python.lint.isort/isort.lock
13:03:27  13:03:26.62 [INFO] Completed: Building flake8.pex from resource://pants.backend.python.lint.flake8/flake8.lock
13:03:27  13:03:26.63 [INFO] Completed: Building autoflake.pex from resource://pants.backend.python.lint.autoflake/autoflake.lock
13:03:27  13:03:26.67 [INFO] Completed: Building docformatter.pex from resource://pants.backend.python.lint.docformatter/docformatter.lock
13:03:27  13:03:26.85 [INFO] Completed: Building black.pex from resource://pants.backend.python.lint.black/black.lock
13:03:27  13:03:26.85 [INFO] Completed: Format with docformatter - docformatter made no changes.
13:03:27  13:03:26.86 [INFO] Completed: Format with isort - isort made no changes.
13:03:27  13:03:26.93 [INFO] Completed: Fix with Autoflake - autoflake made no changes.
13:03:27  13:03:26.97 [INFO] Completed: Lint with Flake8 - flake8 succeeded.
13:03:27  Partition: ['CPython==3.9.*']
13:03:27  
13:03:27  
13:03:27  13:03:27.26 [INFO] Completed: Format with Black - black made no changes.
13:03:27  
13:03:27  ✓ autoflake succeeded.
13:03:27  ✓ black succeeded.
13:03:27  ✓ docformatter succeeded.
13:03:27  ✓ flake8 succeeded.
13:03:27  ✓ isort succeeded.
13:03:27  13:03:27.51 [INFO] Starting: Building mypy.pex from resource://pants.backend.python.typecheck.mypy/mypy.lock
13:03:27  13:03:27.57 [INFO] Starting: Building 4 requirements for requirements.pex from the 3rdparty/default.lock resolve: click, pantsbuild.pants.testutil==2.21.0, pantsbuild.pants==2.21.0, pytest
13:03:29  13:03:29.32 [INFO] Completed: Building mypy.pex from resource://pants.backend.python.typecheck.mypy/mypy.lock
13:03:31  13:03:31.13 [INFO] Completed: Building 4 requirements for requirements.pex from the 3rdparty/default.lock resolve: click, pantsbuild.pants.testutil==2.21.0, pantsbuild.pants==2.21.0, pytest
13:03:31  13:03:31.18 [INFO] Starting: Building requirements_venv.pex
13:03:32  13:03:32.72 [INFO] Completed: Building requirements_venv.pex
13:03:34  13:03:34.63 [INFO] Completed: Typecheck using MyPy - mypy - mypy succeeded.
13:03:34  Success: no issues found in 9 source files
13:03:34  
13:03:34  
13:03:34  
13:03:34  ✓ mypy succeeded.
13:03:34  13:03:34.74 [INFO] Starting: Building pytest.pex from resource://pants.backend.python.subsystems/pytest.lock
13:03:34  13:03:34.77 [INFO] Starting: Building 3 requirements for requirements.pex from the 3rdparty/default.lock resolve: pantsbuild.pants.testutil==2.21.0, pantsbuild.pants==2.21.0, pytest
13:03:34  13:03:34.77 [INFO] Starting: Building requirements.pex
13:03:34  13:03:34.86 [INFO] Canceled: Building pytest.pex from resource://pants.backend.python.subsystems/pytest.lock
13:03:34  13:03:34.86 [INFO] Starting: Building pytest.pex from resource://pants.backend.python.subsystems/pytest.lock
13:03:36  13:03:36.38 [INFO] Completed: Building requirements.pex
13:03:36  13:03:36.52 [INFO] Completed: Building pytest.pex from resource://pants.backend.python.subsystems/pytest.lock
13:03:36  13:03:36.55 [INFO] Starting: Building local_dists.pex
13:03:37  13:03:37.55 [INFO] Completed: Building 3 requirements for requirements.pex from the 3rdparty/default.lock resolve: pantsbuild.pants.testutil==2.21.0, pantsbuild.pants==2.21.0, pytest
13:03:38  13:03:38.05 [INFO] Completed: Building local_dists.pex
13:03:38  13:03:38.10 [INFO] Starting: Building pytest_runner.pex
13:03:38  13:03:38.10 [INFO] Starting: Building pytest_runner.pex
13:03:39  13:03:39.64 [INFO] Completed: Building pytest_runner.pex
13:03:39  13:03:39.73 [INFO] Completed: Building pytest_runner.pex
13:03:40  13:03:40.09 [INFO] Completed: Run Pytest - src/.../x:tests - succeeded.
13:03:42  13:03:41.70 [INFO] Completed: Run Pytest - src/.../y:tests - succeeded.
13:03:42  
13:03:42  ✓ src/.../x:tests succeeded in 0.43s.
13:03:42  ✓ src/.../y:tests succeeded in 1.91s.

# --vvvvv--  Package begins here  --vvvvv--

13:03:42  13:03:41.82 [INFO] Starting: Building 1 requirement for x.pex from the 3rdparty/default.lock resolve: click
13:03:42  13:03:41.86 [INFO] Starting: Building build_backend.pex from resource://pants.backend.python.subsystems/setuptools.lock
13:03:43  13:03:43.30 [INFO] Completed: Building 1 requirement for x.pex from the 3rdparty/default.lock resolve: click
13:03:43  13:03:43.31 [ERROR] 1 Exception encountered:
13:03:43  
13:03:43  Engine traceback:
13:03:43    in `package` goal
13:03:43  
13:03:43  ProcessExecutionFailure: Process 'Building 1 requirement for x.pex from the 3rdparty/default.lock resolve: click' failed with exit code 102.
13:03:43  stdout:
13:03:43  
13:03:43  stderr:
13:03:43  Could not find a compatible interpreter.
13:03:43  
13:03:43  Examined the following interpreters:
13:03:43  1.) /usr/bin/python2.7 CPython==2.7.18
13:03:43  2.) /usr/bin/python3.8 CPython==3.8.10
13:03:43  
13:03:43  No interpreter compatible with the requested constraints was found:
13:03:43  
13:03:43    Version matches CPython==3.9.*
13:03:43  
13:03:43  
13:03:43  
13:03:43  Use `--keep-sandboxes=on_failure` to preserve the process chroot for inspection.
```

--- a/docs/notes/2.27.x.md
+++ b/docs/notes/2.27.x.md
@@ -50,6 +50,8 @@ In [the `[ruff]` subsystem](https://www.pantsbuild.org/2.27/reference/subsystems
 
 The Python Build Standalone backend (`pants.backend.python.providers.experimental.python_build_standalone`) has release metadata current through PBS release `20250317`.
 
+Packaging a `pex_binary` target for a local interpreter (that is, without `complete_platforms`) now works when using a Python provider like `pants.backend.python.providers.experimental.python_build_standalone` or `pants.backend.python.providers.experimental.pyenv`.
+
 The default module mappings now includes the `hdrhistogram` package (imported as `hdrh`).
 
 Minor fixes:

--- a/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
+++ b/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
@@ -20,6 +20,7 @@ from pants.backend.python.goals.package_pex_binary import (
     PexFromTargetsRequestForBuiltPackage,
 )
 from pants.backend.python.macros.python_artifact import PythonArtifact
+from pants.backend.python.providers.python_build_standalone import rules as pbs
 from pants.backend.python.subsystems.setuptools import PythonDistributionFieldSet
 from pants.backend.python.target_types import (
     PexBinary,
@@ -38,7 +39,6 @@ from pants.core.target_types import (
     ResourcesGeneratorTarget,
 )
 from pants.core.target_types import rules as core_target_types_rules
-from pants.testutil.pants_integration_test import run_pants, setup_tmpdir
 from pants.testutil.python_interpreter_selection import skip_unless_python38_present
 from pants.testutil.python_rule_runner import PythonRuleRunner
 from pants.testutil.rule_runner import QueryRule
@@ -465,27 +465,40 @@ def test_package_with_python_provider() -> None:
     # binary when using a Python that isn't automatically visible on $PATH (using the PBS provider
     # as just one way to get such a Python)
 
-    sources = {
-        "app.py": "",
-        "BUILD": dedent(
-            """\
-            python_sources()
-            pex_binary(name="target", entry_point="./app.py")
-            """
-        ),
-    }
+    rule_runner = PythonRuleRunner(
+        rules=[
+            *package_pex_binary.rules(),
+            *pex_from_targets.rules(),
+            *target_types_rules.rules(),
+            *core_target_types_rules(),
+            *pbs.rules(),
+            QueryRule(BuiltPackage, [PexBinaryFieldSet]),
+        ],
+        target_types=[
+            PexBinary,
+            PythonSourcesGeneratorTarget,
+        ],
+    )
 
-    with setup_tmpdir(sources) as tmpdir:
-        run_pants(
-            [
-                "--backend-packages=pants.backend.python",
-                "--backend-packages=pants.backend.python.providers.experimental.python_build_standalone",
-                # a random (https://xkcd.com/221/) old version of Python, that seems unlikely to be
-                # installed on most systems, by default
-                "--python-interpreter-constraints=CPython==3.10.2",
-                "package",
-                f"{tmpdir}:target",
-            ],
-        ).assert_success()
+    rule_runner.write_files(
+        {
+            "app.py": "",
+            "BUILD": dedent(
+                """\
+                python_sources(name="src")
+                pex_binary(name="target", entry_point="./app.py")
+                """
+            ),
+        }
+    )
 
-        assert os.path.isfile(f"dist/{tmpdir}/target.pex")
+    tgt = rule_runner.get_target(Address("", target_name="target"))
+    field_set = PexBinaryFieldSet.create(tgt)
+
+    # a random (https://xkcd.com/221/) old version of Python, that seems unlikely to be installed on
+    # most systems, by default
+    rule_runner.set_options(["--python-interpreter-constraints=CPython==3.10.2"])
+
+    result = rule_runner.request(BuiltPackage, [field_set])
+    assert len(result.artifacts) == 1
+    assert result.artifacts[0].relpath == "target.pex"

--- a/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
+++ b/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
@@ -480,7 +480,8 @@ def test_package_with_python_provider() -> None:
             [
                 "--backend-packages=pants.backend.python",
                 "--backend-packages=pants.backend.python.providers.experimental.python_build_standalone",
-                # a "random" old version of Python, that seems unlikely to be installed on most systems, by default
+                # a random (https://xkcd.com/221/) old version of Python, that seems unlikely to be
+                # installed on most systems, by default
                 "--python-interpreter-constraints=CPython==3.10.2",
                 "package",
                 f"{tmpdir}:target",

--- a/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
+++ b/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
@@ -495,7 +495,10 @@ def test_package_with_python_provider() -> None:
     field_set = PexBinaryFieldSet.create(tgt)
 
     # a random (https://xkcd.com/221/) old version of Python, that seems unlikely to be installed on
-    # most systems, by default
+    # most systems, by default... but also, we don't propagate PATH (etc.) for this rule_runner, so
+    # the test shouldn't be able to find system interpreters anyway.
+    #
+    # (Thus we have two layers of "assurance" the test is doing what is intended.)
     rule_runner.set_options(["--python-interpreter-constraints=CPython==3.10.2"])
 
     result = rule_runner.request(BuiltPackage, [field_set])

--- a/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
+++ b/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
@@ -4,6 +4,7 @@
 from __future__ import annotations
 
 import json
+import os
 import os.path
 import pkgutil
 import subprocess
@@ -37,6 +38,7 @@ from pants.core.target_types import (
     ResourcesGeneratorTarget,
 )
 from pants.core.target_types import rules as core_target_types_rules
+from pants.testutil.pants_integration_test import run_pants, setup_tmpdir
 from pants.testutil.python_interpreter_selection import skip_unless_python38_present
 from pants.testutil.python_rule_runner import PythonRuleRunner
 from pants.testutil.rule_runner import QueryRule
@@ -456,3 +458,33 @@ def test_extra_build_args(rule_runner: PythonRuleRunner) -> None:
 
     assert additional_args[-2] == "--example-extra-arg"
     assert additional_args[-1] == "value-goes-here"
+
+
+def test_package_with_python_provider() -> None:
+    # Per https://github.com/pantsbuild/pants/issues/21048, test packaging a local/unconstrained pex
+    # binary when using a Python that isn't automatically visible on $PATH (using the PBS provider
+    # as just one way to get such a Python)
+
+    sources = {
+        "app.py": "",
+        "BUILD": dedent(
+            """\
+            python_sources()
+            pex_binary(name="target", entry_point="./app.py")
+            """
+        ),
+    }
+
+    with setup_tmpdir(sources) as tmpdir:
+        run_pants(
+            [
+                "--backend-packages=pants.backend.python",
+                "--backend-packages=pants.backend.python.providers.experimental.python_build_standalone",
+                # a "random" old version of Python, that seems unlikely to be installed on most systems, by default
+                "--python-interpreter-constraints=CPython==3.10.2",
+                "package",
+                f"{tmpdir}:target",
+            ],
+        ).assert_success()
+
+        assert os.path.isfile(f"dist/{tmpdir}/target.pex")

--- a/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
+++ b/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
@@ -4,7 +4,6 @@
 from __future__ import annotations
 
 import json
-import os
 import os.path
 import pkgutil
 import subprocess

--- a/src/python/pants/backend/python/util_rules/pex.py
+++ b/src/python/pants/backend/python/util_rules/pex.py
@@ -464,7 +464,7 @@ async def _determine_pex_python_and_platforms(request: PexRequest) -> _BuildPexP
             [
                 *request.interpreter_constraints.generate_pex_arg_list(),
                 "--python-path",
-                os.path.dirname(python.path),
+                python.path,
             ],
         )
 

--- a/src/python/pants/backend/python/util_rules/pex.py
+++ b/src/python/pants/backend/python/util_rules/pex.py
@@ -456,7 +456,17 @@ async def _determine_pex_python_and_platforms(request: PexRequest) -> _BuildPexP
     else:
         # `--interpreter-constraint` options are mutually exclusive with the `--python` option,
         # so we only specify them if we have not already located a concrete Python.
-        return _BuildPexPythonSetup(None, request.interpreter_constraints.generate_pex_arg_list())
+        python = await Get(
+            PythonExecutable, InterpreterConstraints, request.interpreter_constraints
+        )
+        return _BuildPexPythonSetup(
+            python,
+            [
+                *request.interpreter_constraints.generate_pex_arg_list(),
+                "--python-path",
+                os.path.dirname(python.path),
+            ],
+        )
 
     return _BuildPexPythonSetup(python, ["--python", python.path])
 

--- a/src/python/pants/backend/python/util_rules/pex.py
+++ b/src/python/pants/backend/python/util_rules/pex.py
@@ -446,19 +446,24 @@ async def _determine_pex_python_and_platforms(request: PexRequest) -> _BuildPexP
 
     if request.python:
         python = request.python
-    elif request.internal_only:
-        # NB: If it's an internal_only PEX, we do our own lookup of the interpreter based on the
-        # interpreter constraints, and then will run the PEX with that specific interpreter. We
-        # will have already validated that there were no platforms.
-        python = await Get(
-            PythonExecutable, InterpreterConstraints, request.interpreter_constraints
-        )
     else:
-        # `--interpreter-constraint` options are mutually exclusive with the `--python` option,
-        # so we only specify them if we have not already located a concrete Python.
         python = await Get(
             PythonExecutable, InterpreterConstraints, request.interpreter_constraints
         )
+
+    if request.python or request.internal_only:
+        # Sometimes we want to build and run with a specific interpreter (either because request
+        # demanded it, or because it's an internal-only PEX). We will have already validated that
+        # there were no platforms.
+        return _BuildPexPythonSetup(python, ["--python", python.path])
+
+    else:
+        # Otherwise, we don't want to force compatibility with a particular interpreter (as in, the
+        # resulting PEX should follow the ICs), but we _do_ want to tell PEX about at least one
+        # interpreter that is compatible, to ensure that an interpreter installed/managed by
+        # provider backends are visible (in the extreme case, a machine may have no Python
+        # interpreters installed at all, and just rely on Pants' provider backends to install them,
+        # and thus pex searching $PATH will find nothing).
         return _BuildPexPythonSetup(
             python,
             [
@@ -467,8 +472,6 @@ async def _determine_pex_python_and_platforms(request: PexRequest) -> _BuildPexP
                 python.path,
             ],
         )
-
-    return _BuildPexPythonSetup(python, ["--python", python.path])
 
 
 @dataclass

--- a/src/python/pants/backend/python/util_rules/pex_test.py
+++ b/src/python/pants/backend/python/util_rules/pex_test.py
@@ -656,10 +656,15 @@ def test_determine_pex_python_and_platforms() -> None:
         )
         assert result == expected
 
-    assert_setup(expected=_BuildPexPythonSetup(None, []))
+    assert_setup(
+        expected=_BuildPexPythonSetup(discovered_python, ["--python-path", discovered_python.path])
+    )
     assert_setup(
         interpreter_constraints=ics,
-        expected=_BuildPexPythonSetup(None, ["--interpreter-constraint", "CPython==3.8"]),
+        expected=_BuildPexPythonSetup(
+            discovered_python,
+            ["--interpreter-constraint", "CPython==3.8", "--python-path", discovered_python.path],
+        ),
     )
     assert_setup(
         internal_only=True,

